File: Numeros Pseudo-Aleatorios/Add With Carry/AddWithCarry.go
/*
 PRNG .......: Add With Carry (AWC)
 Ano ........: 1991
 Autor ......: George Marsaglia
*/

package main

import (
	"fmt"
	"time"
)

func AWCNextRandom(Xn *uint64) uint64 {
	*Xn = (*Xn + 12345) % 4294967296
	return *Xn
}

func AWCInitialize(seed uint64, Xn *uint64) {
	*Xn = seed
}

func main() {

	fmt.Println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
	fmt.Println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
	fmt.Println("PADRAO: Add With Carry")
	fmt.Println("==========================================")
	fmt.Println()
	fmt.Println("    Semente: 1234        Semente: TIME")
	fmt.Println("    -------------        -------------")

	var Xn1, Xn2 uint64
	var X1, X2 uint64

	X1 = 1234
	X2 = uint64(time.Now().Unix())
  
	AWCInitialize(X1, &Xn1)
	AWCInitialize(X2, &Xn2)

	for i := 1; i <= 10; i++ {
		fmt.Printf("%17d        %13d\n", AWCNextRandom(&Xn1), AWCNextRandom(&Xn2))
	}

}

File: Numeros Pseudo-Aleatorios/Add With Carry/AddWithCarry.pas
{ 
 PRNG .......: Add With Carry (AWC)
 Ano ........: 1991
 Autor ......: George Marsaglia
}

program AddWithCarry;

uses
  SysUtils, DateUtils;

var
  Xn1, Xn2: LongWord;

function AWCNextRandom(var Xn: LongWord): LongWord;
begin
  Xn := (Xn + 12345) mod 4294967296;
  AWCNextRandom := Xn;
end;

procedure AWCInitialize(seed: LongWord; var Xn: LongWord);
begin
  Xn := seed;
end;

var
  i: Integer;
  seed1, seed2: LongWord;
  
begin

  writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  writeln('PADRAO: Add With Carry');
  writeln('==========================================');
  writeln;
  writeln('    Semente: 1234        Semente: TIME');
  writeln('    -------------        -------------');

  seed1 := 1234;
  AWCInitialize(seed1, Xn1);

  seed2 := LongWord(DateTimeToUnix(Now));
  AWCInitialize(seed2, Xn2);

  for i := 1 to 10 do
    writeln(AWCNextRandom(Xn1):17, '        ', AWCNextRandom(Xn2):13);
    
end.
